### PR TITLE
chore(dsm): handle when dd header not present for consuming service

### PR DESCRIPTION
**Pull request by shethhriday29**
Previously, if a service receiving SQS messages didn't find a datadog header in the message that included the data streams context, it would log a warning message and not set a checkpoint, thereby not relaying any information to the dsm processor. However, other DSM technologies/runtimes handle the edge case by still setting the pathway to `None` and using this pathway to call `set_checkpoint`.

PR was tested by running all the botocore tests, including the DSM tests, to ensure they still work. No new technical risks are introduced by this PR.

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](../docs/contributing.rst#release-branch-maintenance))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](../docs/contributing.rst#release-branch-maintenance)

### CODE REVIEW

File: ddtrace/contrib/botocore/patch.py [hunk 1]
```diff
@@ -524,27 +524,14 @@ def patched_api_call(original_func, instance, args, kwargs):
...
-                                    "Unable to find trace context for the message. Please ensure that your Datadog "
-                                    "agent is correctly configured on the application sending the message and that "
-                                    "the SQS messages being sent have 9 or less Message Attributes."
+                                pathway = json.loads(message["MessageAttributes"]["_datadog"]["StringValue"]).get(
+                                    PROPAGATION_KEY_BASE_64, None
                                 )
+                            else:
+                                pathway = None
+
+                            ctx = pin.tracer.data_streams_processor.decode_pathway_b64(pathway)
```
Comment (tabgok) [inline ddtrace/contrib/botocore/patch.py:533]: does this need to be added to the MessageAttributes?

Comment (mabdinur) [reply ddtrace/contrib/botocore/patch.py:533]: `decode_pathway_b64()` expects a string. Passing None should raise an Error.

File: releasenotes/notes/botocore_dsm_sqs_no_header-15e92adad81448ea.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+other:
```
Comment (mabdinur) [inline releasenotes/notes/botocore_dsm_sqs_no_header-15e92adad81448ea.yaml:2]: If this is a fix for an unreleased change then a release note is not necessary

**Review by mabdinur — COMMENTED** [review]
Can we add a regression test. Some thing that validates that set_checkpoint sets the expected tags.

File: ddtrace/contrib/botocore/patch.py (hunk 1 above)
Comment (tabgok) [reply ddtrace/contrib/botocore/patch.py:533]: discussed offline, the answer is "no", as we are currently only looking at in/out pairs.

File: ddtrace/contrib/botocore/patch.py [hunk 3]
```diff
@@ -524,27 +524,14 @@ def patched_api_call(original_func, instance, args, kwargs):
...
-                            else:
-                                log.debug(
-                                    "Unable to find trace context for the message. Please ensure that your Datadog "
-                                    "agent is correctly configured on the application sending the message and that "
-                                    "the SQS messages being sent have 9 or less Message Attributes."
+                                pathway = json.loads(message["MessageAttributes"]["_datadog"]["StringValue"]).get(
+                                    PROPAGATION_KEY_BASE_64, None
                                 )
+                            else:
+                                pathway = None
```
Comment (tabgok) [inline ddtrace/contrib/botocore/patch.py:531]: nit: its a little more readable to set this as None before the if/else

Comment (shethhriday29) [reply ddtrace/contrib/botocore/patch.py]: done!

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/contrib/botocore/patch.py [hunk 4]
```diff
@@ -519,32 +519,17 @@ def patched_api_call(original_func, instance, args, kwargs):
...
-                            else:
-                                log.debug(
-                                    "Unable to find trace context for the message. Please ensure that your Datadog "
-                                    "agent is correctly configured on the application sending the message and that "
-                                    "the SQS messages being sent have 9 or less Message Attributes."
+                                pathway = json.loads(message["MessageAttributes"]["_datadog"]["StringValue"]).get(
+                                    PROPAGATION_KEY_BASE_64, None
                                 )
+                            ctx = pin.tracer.data_streams_processor.decode_pathway_b64(pathway)
+                            ctx.set_checkpoint(["direction:in", "topic:" + queue_name, "type:sqs"])
```
Comment (mabdinur) [inline ddtrace/contrib/botocore/patch.py:532]: We should avoid calling ``data.encode("utf-8")`` when data is None. Handling an AttributeError here is less than ideal. 

Instead we can just create a new pathway directly. With this suggestion we can revert handling the ``AttributeError`` in `decode_pathway_b64()`

```suggestion
                               ctx = pin.tracer.data_streams_processor.decode_pathway_b64(pathway)
                            else:
                               ctx = pin.tracer.data_streams_processor.new_pathway()
                            ctx.set_checkpoint(["direction:in", "topic:" + queue_name, "type:sqs"])
```

Feel free to experiment with alternative approaches.

File: ddtrace/contrib/botocore/patch.py [hunk 5]
```diff
@@ -519,32 +519,17 @@ def patched_api_call(original_func, instance, args, kwargs):
                 try:
                     if "Messages" in result:
                         for message in result["Messages"]:
+                            pathway = None
```
Comment (mabdinur) [inline ddtrace/contrib/botocore/patch.py:522]: Should we move `pathway = None` out of the for loop (ie move it to line 521)

File: ddtrace/contrib/botocore/patch.py (hunk 4 above)
Comment (mabdinur) [inline ddtrace/contrib/botocore/patch.py:532]: We should avoid calling data.encode("utf-8") when data is None. Handling an AttributeError here is less than ideal.

Instead we can just create a new pathway directly. With this suggestion we can revert handling the AttributeError in decode_pathway_b64()

```suggestion
                            if pathway is not None:
                               ctx = pin.tracer.data_streams_processor.decode_pathway_b64(pathway)
                            else:
                               ctx = pin.tracer.data_streams_processor.new_pathway()
                            ctx.set_checkpoint(["direction:in", "topic:" + queue_name, "type:sqs"])
```

File: ddtrace/internal/datastreams/processor.py [hunk 6]
```diff
@@ -262,10 +262,13 @@ def decode_pathway(self, data):
...
-        binary_pathway = data.encode("utf-8")
-        encoded_pathway = base64.b64decode(binary_pathway)
-        data_streams_context = self.decode_pathway(encoded_pathway)
-        return data_streams_context
+        try:
+            binary_pathway = data.encode("utf-8")
+            encoded_pathway = base64.b64decode(binary_pathway)
+            data_streams_context = self.decode_pathway(encoded_pathway)
+            return data_streams_context
+        except AttributeError:
```
Comment (mabdinur) [inline ddtrace/internal/datastreams/processor.py:270]: If the change above is accepted we can revert this or at least we should change the method signature to `type: Optional[str] -> DataStreamsCtx ` and then handle the case were data is None without raise an Exception.

File: ddtrace/contrib/botocore/patch.py (hunk 5 above)
Comment (shethhriday29) [reply ddtrace/contrib/botocore/patch.py:522]: we need to reset `pathway` to `None` for every message, we could have a case where some messages being received may have the `PROPAGATION_KEY_BASE_64` and others may not, and for messages that don't, we don't want other messages' `PROPAGATION_KEY_BASE_64` being used (rather just `None`)

**Review by mabdinur — APPROVED** [review]
